unix/w0vncserver/wlr/WlrDesktop.cxx

```diff
@@ -0,0 +1,115 @@
+/* Copyright 2025 Adam Halim for Cendio AB
+ *
+ * This is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <assert.h>
+#include <unistd.h>
+#include <sys/mman.h>
+
+#include <glib.h>
+#include <wayland-client.h>
+#include <wayland-client-core.h>
+#include <wayland-client-protocol.h>
+
+#include <core/LogWriter.h>
+#include <rfb/VNCServerST.h>
+
+#include "../w0vncserver.h"
+#include "../wayland/GWaylandSource.h"
+#include "../wayland/WDisplay.h"
+#include "../wayland/WOutput.h"
+#include "../wayland/WShm.h"
+#include "WlrPixelBuffer.h"
+#include "WlrDesktop.h"
+
+static core::LogWriter vlog("WlrDesktop");
+
+WlrDesktop::WlrDesktop(GMainLoop* loop_)
+  : server(nullptr), loop(loop_), display(nullptr),
+    seat(nullptr), pb(nullptr), wlrSource(nullptr)
+{
+  assert(available());
+
+  display = new WDisplay();
+  output = new WOutput(display);
+}
+
+WlrDesktop::~WlrDesktop()
+{
+  delete pb;
+  delete wlrSource;
+  delete output;
+  delete display;
+}
+
+void WlrDesktop::init(rfb::VNCServer* vs)
+{
+  server = vs;
+}
+
+void WlrDesktop::start()
+{
+  std::function<void()> desktopReadyCb = [this]() {
+    server->setPixelBuffer(pb);
+  };
+
+  pb = new WlrPixelBuffer(display, output, server, desktopReadyCb);
+
+  wlrSource = new GWaylandSource(display);
+  wlrSource->attach(g_main_loop_get_context(loop));
+}
+
+void WlrDesktop::stop()
+{
+  server->setPixelBuffer(nullptr);
+
+  delete wlrSource;
+  wlrSource = nullptr;
+
+  delete pb;
+  pb = nullptr;
+}
+
+void WlrDesktop::frameTick(uint64_t /* msc */)
+{
+  // FIXME: Should we use the monitor refresh rate instead?
+  pb->captureFrame();
```

Comment:
This is a bit backwards for how the system is designed to work. The desktop should be pushing changes to the server, not the other way around.

Can we get a notification when something changes instead? And if reading is optional, it would be nice if we can delay reading until needed. Not every screen updated will be sent to client.